library/experiments/hate_speech_multi_type_hf.py

```diff
@@ -73,38 +73,58 @@
 
 """ Data """
 
-input_data = DataSource("input")
-
-dataloader = get_tweet_eval_dataloader("hate")
+tweet_eval_hate = DataSource("tweet_eval_hate", get_tweet_eval_dataloader("hate"))
+dynahate = DataSource("dynahate", get_dynahate_dataloader())
 
 
 """ Pipelines"""
 
 hf_distilbert = Pipeline(
     "distilbert-binary",
-    input_data,
-    [
-        HuggingfaceModel("distilbert-binary", huggingface_distil_bert_config),
+    datasource_fit=tweet_eval_hate,
+    datasource_predict=tweet_eval_hate,
+    models=[
+        HuggingfaceModel(
+            "distilbert-binary",
+            huggingface_distil_bert_config,
+            dict_lookup={"LABEL_0": 0, "LABEL_1": 1},
+        ),
         ClassificationOutputAdaptor(select=0),
     ],
 )
 
 hf_distilbert_uncased_emotion = Pipeline(
     "distilbert-emotion",
-    input_data,
-    [
+    datasource_fit=dynahate,
+    datasource_predict=tweet_eval_hate,
+    models=[
         HuggingfaceModel(
-            "distilbert-emotion", huggingface_distilbert_uncased_emotion_config
+            "distilbert-emotion",
+            huggingface_distilbert_uncased_emotion_config,
+            dict_lookup={
+                "sadness": 0,
+                "joy": 1,
+                "anger": 2,
+                "fear": 3,
+                "love": 4,
+                "surprise": 5,
+            },
         ),
         ClassificationOutputAdaptor(select=0),
     ],
 )
 
+vector_concat = VectorConcat(
+    "concat-source",
+    [hf_distilbert, hf_distilbert_uncased_emotion],
+    datasource_labels=tweet_eval_hate,
+)
 
 full_pipeline = Pipeline(
     "nlp_hf_meta-model-pipeline",
-    VectorConcat("concat-source", [hf_distilbert, hf_distilbert_uncased_emotion]),
-    [
+    datasource_fit=vector_concat,
```

Comment:
should we make one of these optional? I'm wondering because repeating dataloaders two times for 98% of the cases is probably an overkill?

Reply:
Yes, I agree! I wanted to first do it without it being optional and then figure out an interface to only pass in one.

Options
- [ ] ``datasource_predict`` is optional and if not given it defaults to ``datasource_fit``
- [ ] ``datasource_fit`` is optional and if not given it defaults to ``datasource_predict``
- [ ] We add another variable named simply ``datasource`` as a third option

Reply:
if we keep the splits at the experiment level (do we?), then I'd chose the first option, "datasource_predict being optional"